src/main/java/org/apache/spark/shuffle/sort/RssShuffleExternalSorter.java

```diff
@@ -0,0 +1,411 @@
+                        } else {
+                            prevPartition = partition;
+                        }
+                        currentPartition = partition;
+
+                        final long recordPointer = sortedRecords.packedRecordPointer.getRecordPointer();
+                        recordPage = taskMemoryManager.getPage(recordPointer);
+                        final long recordOffsetInPage = taskMemoryManager.getOffsetInPage(recordPointer);
+                        currentPartitionDataRemaining = UnsafeAlignedOffset.getSize(recordPage, recordOffsetInPage);
+                        recordReadPosition = recordOffsetInPage + uaoSize; // skip over record length
+                        return true;
+                    } else {
+                        return false;
+                    }
+                } else {
+                    return true;
+                }
+            }
+
+            @Override
+            public Tuple3<Object, byte[], Integer> next() {
+
+                boolean condition = true;
+                int partitionIdToSpill = currentPartition;
+                arrayFillPosition = 0;
+
+                while (condition) {
+
+                    final int toTransfer = Math.min((int) writeBufferSize - arrayFillPosition, currentPartitionDataRemaining);
+
+                    // ToDo: This is an issue when the records are too large
+                    if (toTransfer < currentPartitionDataRemaining) {
+                        condition = false;
+                    } else {
+                        Platform.copyMemory(recordPage, recordReadPosition, byteArr,
+                                Platform.BYTE_ARRAY_OFFSET + arrayFillPosition, toTransfer);
+                        recordReadPosition += toTransfer;
+                        currentPartitionDataRemaining -= toTransfer;
+                        bytesWritten += toTransfer;
+                        arrayFillPosition += toTransfer;
+
+                        if (arrayFillPosition >= writeBufferSize) {
+                            // There is no more space left in byteArr
+                            assert (arrayFillPosition == byteArr.length);
+                            assert (arrayFillPosition == writeBufferSize);
+                            partitionIdToSpill = currentPartition;
+                            condition = false;
+                        } else if (hasNext()) {
+                            if (currentPartition != prevPartition) {
+                                // byteArr has more space but there no data for the previous partition anymore
+                                partitionIdToSpill = prevPartition;
+                                condition = false;
+                            } else {
+                                // byteArr has more space and there is more data in the current partition,
+                                // lets fetch more
+                                condition = true;
```

Comment:
what about the case when this record+ current bytarray.size > bytearray total size